[android] fix memory leak in `TabbedPage`

jonathanpeppers opened the pull request:
Fixes: https://github.com/dotnet/maui/issues/17959
Context: https://github.com/VoznichkaSviatoslav1/TabbedPageMemoryLeak

In the sample app, I added some `GC.Collect()` calls to force the GC to run, but the following log message never prints:

    public partial class MyTabbedPage : TabbedPage
    {
        public MyTabbedPage()
        {
            InitializeComponent();

            Console.WriteLine("Constructor was called");
        }

        // Doesn't print!!!
        ~MyTabbedPage() => Console.WriteLine("Destructor was called");

I took a memory snapshot with `dotnet-gcdump` and found a tree like:

    SampleApp.MyTabbedPage ->
        Microsoft.Maui.Controls.Platform.MultiPageFragmentStateAdapter<Microsoft.Maui.Controls.Page> ->
            Action<Microsoft.Maui.Controls.Platform.AdapterItemKey>

I didn't find an explanation of what was holding
`MultiPageFragmentStateAdapter`. I tried eliminating the `Action<T>`, but that didn't solve the underlying issue; it just made the object tree shorter:

    SampleApp.MyTabbedPage ->
        Microsoft.Maui.Controls.Platform.MultiPageFragmentStateAdapter<Microsoft.Maui.Controls.Page>

I could also reproduce the problem in a new device test.

Reviewing the code, I found we were doing:

    _viewPager.Adapter = new MultiPageFragmentStateAdapter<Page>(tabbedPage, FragmentManager, _context) { CountOverride = tabbedPage.Children.Count };

But then in the cleanup code, we never set `Adapter` to `null`:

    _viewPager.Adapter = null;

After this change, the memory is released as expected. I am not exactly sure *why* this leaked, but the change does seem appropriate and fixes the problem.

Now the new device test passes & the sample app prints:

    03-14 15:06:32.394  6055  6055 I DOTNET  : Constructor was called
    03-14 15:06:37.008  6055  6089 I DOTNET  : Destructor was called

jonathanpeppers committed:
[android] fix memory leak in `TabbedPage`

Fixes: https://github.com/dotnet/maui/issues/17959
Context: https://github.com/VoznichkaSviatoslav1/TabbedPageMemoryLeak

In the sample app, I added some `GC.Collect()` calls to force the GC
to run, but the following log message never prints:

    public partial class MyTabbedPage : TabbedPage
    {
        public MyTabbedPage()
        {
            InitializeComponent();

            Console.WriteLine("Constructor was called");
        }

        // Doesn't print!!!
        ~MyTabbedPage() => Console.WriteLine("Destructor was called");

I took a memory snapshot with `dotnet-gcdump` and found a tree like:

    SampleApp.MyTabbedPage ->
        Microsoft.Maui.Controls.Platform.MultiPageFragmentStateAdapter<Microsoft.Maui.Controls.Page> ->
            Action<Microsoft.Maui.Controls.Platform.AdapterItemKey>

I didn't find an explanation of what was holding
`MultiPageFragmentStateAdapter`. I tried eliminating the `Action<T>`,
but that didn't solve the underlying issue; it just made the object
tree shorter:

    SampleApp.MyTabbedPage ->
        Microsoft.Maui.Controls.Platform.MultiPageFragmentStateAdapter<Microsoft.Maui.Controls.Page>

I could also reproduce the problem in a new device test.

Reviewing the code, I found we were doing:

    _viewPager.Adapter = new MultiPageFragmentStateAdapter<Page>(tabbedPage, FragmentManager, _context) { CountOverride = tabbedPage.Children.Count };

But then in the cleanup code, we never set `Adapter` to `null`:

    _viewPager.Adapter = null;

After this change, the memory is released as expected. I am not
exactly sure *why* this leaked, but the change does seem appropriate
and fixes the problem.

Now the new device test passes & the sample app prints:

    03-14 15:06:32.394  6055  6055 I DOTNET  : Constructor was called
    03-14 15:06:37.008  6055  6089 I DOTNET  : Destructor was called

jonathanpeppers commented:
So I fixed Android, but the new test discovered it leaks on Windows?

![image](https://github.com/dotnet/maui/assets/840039/d0eb308d-98a4-40fc-b726-d38d63adb254)

jsuarezruiz commented:
Weird, I have triggered the UITests again.
![image](https://github.com/dotnet/maui/assets/6755973/71f04a4a-351a-44db-8242-1a65bfd156b9)

jonathanpeppers commented:
No Windows is actually leaking when you use `TabbedPage`:

![image](https://github.com/dotnet/maui/assets/840039/e502be55-bf54-47f3-aefd-ee9d6bdfe395)

I haven't figured it out yet.

jonathanpeppers committed:
[windows] skip test

I spent 2.5 hours on this, and did not figure it out! Comment for now as original issue is Android.

jonathanpeppers commented:
I disabled the test on Windows for now, since the original issue is Android. I spent 2 hours commenting out code, and can't seem to figure it out.

I don't understand what the `[Dependent Handle]` entries in the above screenshot are on Windows. Is there a doc somewhere on internals of WindowsAppSDK?

jonathanpeppers committed:
Somehow duplicate [Fact]
